### PR TITLE
Allow `__init__` and `__new__` to return `NoReturn`

### PR DESCRIPTION
**Pull request by BvB93**
Closes https://github.com/python/mypy/issues/10342

This PR allows `__init__` and `__new__` methods to return `NoReturn`, allowing for the annotation of classes
that cannot be directly initialized (_e.g._ those that are defined in C such as `types.GeneratorType`).

### CODE REVIEW

File: test-data/unit/check-classes.test
Comment (BvB93) [inline test-data/unit/check-classes.test:6910]:
``` python
class E:
    def __new__(cls) -> NoReturn: ...
    def __init__(self) -> None: ...
```
Notable in it's absence is the example above. Unfortunately this example fails to properly recognize `NoReturn` when instantiating the class, as mypy tends to ignore `__new__` if `__init__` is defined:
https://github.com/python/mypy/blob/8e82171a3e68a5180fab267cad4d2b7cfa1f5cdc/mypy/checkmember.py#L969-L973

File: test-data/unit/check-classes.test
Comment (BvB93) [reply test-data/unit/check-classes.test]: Fair enough; both suggestions have been implemented as of https://github.com/python/mypy/pull/11216/commits/73948b1aa125114837defd34bd39d3dc10893382.

**Comment by BvB93** [comment]
Any idea what's going on with the failed `Run mypy_primer / Run (2)` Test?
It looks like it crashed while installing depencies or something.

**Comment by sobolevn** [comment]
@BvB93 try to close / open this PR. It should be fixed now 👍

**Comment by BvB93** [comment]
Let's give it a shot (note: I might need approval again as a first-time contributor)

**Comment by sobolevn** [comment]
@BvB93 sorry, I cannot approve it 😞 
We need to wait for maintainers to do that.

**Comment by JelleZijlstra** [comment]
I pushed the approval button.

**Comment by benkehoe** [comment]
Today I discovered that I can't annotate a class as unable to be instantiated through `__init__() -> NoReturn`. I'd love to see this PR merged!

**Comment by PedanticHacker** [comment]
Please let me get something straight. So this PR proposes to be able to annotate the return type of a class' `__init__()` with `NoReturn` rather than `None`?

So like
```python
class MyExampleClass:
    def __init__(self) -> NoReturn:
        pass
```

rather than
```python
class MyExampleClass:
    def __init__(self) -> None:
        pass
```

If that's the case, I am very much welcoming the change. It is more understandable to read that a class' `__init__` doesn't return anything (`NoReturn`) as oppose to read that it returns `None` (which techically is not true).

**Comment by BvB93** [comment]
Closing and re-opening as no CI runs appear to be starting.

**Comment by BvB93** [comment]
> Closing and re-opening as no CI runs appear to be starting.

Right, that's better. This PR does still need approval after the latest `master` branch rebase.

> Please let me get something straight. So this PR proposes to be able to annotate the return type of a class' __init__() with NoReturn rather than None?

@PedanticHacker not quite, [`NoReturn`](https://docs.python.org/3/library/typing.html#typing.NoReturn) is used as return-annotation for function that _always_ raise an exception. In the examples you provided `None` is still the appropriate return type, as python functions without an explicit return statement alwways return `None` (note that this is actual runtime behavior; not just a convention).

``` python
class MyExampleClass:
    def __init__(self) -> NoReturn:
        raise TypeError("Cannot initialize 'MyExampleClass'")
```

**Comment by BvB93** [comment]
Rebased to get rid of another merge conflict. For future reference, is there a preference for either rebasing or updating from master when dealing with merge conflicts?

**Comment by JelleZijlstra** [comment]
> Rebased to get rid of another merge conflict. For future reference, is there a preference for either rebasing or updating from master when dealing with merge conflicts?

Merging is generally preferred. When you force-push, the review history for comments before the force-push gets lost in the UI, making it harder for reviewers to review previous changes.

**Review by hauntsaninja — APPROVED** [review]
This looks great! If you fix the merge conflicts and ping me, I will merge :-)

**Comment by iyanging** [comment]
Hi @BvB93 @hauntsaninja, I made another PR which merged master, maybe it's worth to have a look.

**Comment by hauntsaninja** [comment]
This got merged in #13480

**Comment by hauntsaninja** [comment]
Thank you BvB93!

**Pull request closed without merging**
